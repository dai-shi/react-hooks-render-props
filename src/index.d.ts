import * as React from 'react';

// tslint:disable-next-line:no-any
export const useRenderProps: <P extends {}>(c: React.ComponentType<P>, p?: P) => any[];
export const wrap: <P extends {}>(c: React.SFC<P>) => React.SFC<P>;
