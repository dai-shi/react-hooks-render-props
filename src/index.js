import {
  createElement,
  useState,
  useRef,
  useEffect,
} from 'react';

const wrappers = [];

export const useRenderProps = (WrapperComponent, wrapperProps) => {
  const [args, setArgs] = useState([]);
  const ref = useRef({});
  if (!ref.current.initialized) {
    wrappers.push({
      WrapperComponent,
      wrapperProps,
      setArgs,
    });
  }
  useEffect(() => {
    ref.current.initialized = true;
  }, []);
  return args;
};

export const wrap = FunctionComponent => (props) => {
  const element = FunctionComponent(props);
  const ref = useRef({ wrapper: wrappers.pop() });
  const {
    WrapperComponent,
    wrapperProps,
  } = ref.current.wrapper;
  return createElement(WrapperComponent, wrapperProps, (...args) => {
    if (!ref.current.processed) {
      ref.current.wrapper.setArgs(args);
      ref.current.processed = true;
    } else {
      ref.current.processed = false;
    }
    return element;
  });
};
