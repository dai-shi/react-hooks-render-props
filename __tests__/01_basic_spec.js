/* eslint-env jest */

import React from 'react';
import { configure, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import { useRenderProps, wrap } from '../src/index';

configure({ adapter: new Adapter() });

describe('basic spec', () => {
  it('should have a function', () => {
    expect(useRenderProps).toBeDefined();
    expect(wrap).toBeDefined();
  });

  it('should create the same component', () => {
    const SetNumber = ({ children }) => (
      <div>
        {children(123)}
      </div>
    );
    let wrapper1;
    let wrapper2;
    (() => {
      const Component = () => (
        <SetNumber>
          {value => <span>{value}</span>}
        </SetNumber>
      );
      const App = () => (
        <Component />
      );
      wrapper1 = mount(<App />);
    })();
    (() => {
      const Component = wrap(() => {
        const [value] = useRenderProps(SetNumber);
        return (
          <span>{value}</span>
        );
      });
      const App = () => (
        <Component />
      );
      wrapper2 = mount(<App />);
    })();
    expect(wrapper2.html()).toEqual(wrapper1.html());
    expect(toJson(wrapper2)).toMatchSnapshot();
  });
});
