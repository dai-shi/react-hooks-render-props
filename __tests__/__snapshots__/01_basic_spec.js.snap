// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic spec should create the same component 1`] = `
<App>
  <Component>
    <SetNumber>
      <div>
        <span>
          123
        </span>
      </div>
    </SetNumber>
  </Component>
</App>
`;
